configfile: "config/config.yaml"

RESULTDIR = config["outdir"]
ALIGNDIR = config["aligndir"]
LOGDIR = config["logdir"]
SCRIPTDIR = config["scriptdir"]

include: "rules/align.smk"
include: "rules/depth.smk"
include: "rules/subsample.smk"
include: "rules/cutesv.smk"
include: "rules/sniffles.smk"
include: "rules/svim.smk"
include: "rules/pbsv.smk"
include: "rules/npinv.smk"
include: "rules/nanosv.smk"

rule na24385:
    input:
        expand(f"{RESULTDIR}/{{aligner}}/cutesv/{{coverage}}/GM24385.vcf", aligner=["minimap2","ngmlr"], coverage=["5X", "10X", "15X", "20X", "25X", "35X", "total"]),
        expand(f"{RESULTDIR}/{{aligner}}/sniffles/{{coverage}}/GM24385.vcf", aligner=["minimap2","ngmlr"], coverage=["5X", "10X", "15X", "20X", "25X", "35X", "total"]),
        expand(f"{RESULTDIR}/{{aligner}}/svim/{{coverage}}/GM24385.vcf", aligner=["minimap2","ngmlr"], coverage=["5X", "10X", "15X", "20X", "25X", "35X", "total"]),
        expand(f"{RESULTDIR}/pbmm2/pbsv/{{coverage}}/GM24385.vcf",coverage=["5X", "10X", "15X", "20X", "25X", "35X", "total"]),
        expand(f"{RESULTDIR}/{{aligner}}/npinv/{{coverage}}/GM24385.vcf", aligner=["minimap2","ngmlr"], coverage=["5X", "10X", "15X", "20X", "25X", "35X", "total"]),
        expand(f"{RESULTDIR}/{{aligner}}/nanosv/{{coverage}}/GM24385.vcf", aligner=["minimap2","ngmlr"], coverage=["5X", "10X", "15X", "20X", "25X", "35X", "total"])

#first part: map reads to reference and perform kind of QC (coverage)

rule align:
    input:
        #expand(f"{ALIGNDIR}/GM24385.{{aligner}}.srt.bam",aligner=["minimap2","pbmm2","ngmlr"]),
        expand(f"{ALIGNDIR}/GM24385.{{aligner}}.srt.bam.bai",aligner=["minimap2","pbmm2","ngmlr"])

rule depth:
    input:
        #expand(f"{ALIGNDIR}/GM24385.{{aligner}}.mosdepth.global.dist.txt",aligner=["minimap2","pbmm2","ngmlr"]),
        #expand(f"{ALIGNDIR}/GM24385.{{aligner}}.mosdepth.summary.txt",aligner=["minimap2","pbmm2","ngmlr"]),
        #expand(f"{ALIGNDIR}/GM24385.{{aligner}}.mosdepth.region.dist.txt",aligner=["minimap2","pbmm2","ngmlr"]),
        #expand(f"{ALIGNDIR}/GM24385.{{aligner}}.regions.bed.gz",aligner=["minimap2","pbmm2","ngmlr"]),
        #expand(f"{ALIGNDIR}/GM24385.{{aligner}}.regions.bed.gz.csi",aligner=["minimap2","pbmm2","ngmlr"]),
        expand(f"{ALIGNDIR}/GM24385.{{aligner}}.coverage_per_chromosome.png", aligner=["minimap2","pbmm2","ngmlr"]),
        f"{ALIGNDIR}/GM24385.fraction_of_bases_per_depth.png"

rule subsample:
    input:
        #expand(f"{ALIGNDIR}/{{coverage}}/GM24385.{{aligner}}.srt.bam", coverage=["5X", "10X", "15X", "20X", "25X", "35X"], aligner=["minimap2","pbmm2","ngmlr"]),
        expand(f"{ALIGNDIR}/{{coverage}}/GM24385.{{aligner}}.srt.bam.bai", coverage=["5X", "10X", "15X", "20X", "25X", "35X"], aligner=["minimap2","pbmm2","ngmlr"])

#second part: call variants

rule cutesv:
    input:
        expand(f"{RESULTDIR}/{{aligner}}/cutesv/{{coverage}}/GM24385.vcf", aligner=["minimap2","ngmlr"], coverage=["5X", "10X", "15X", "20X", "25X", "35X", "total"])

rule sniffles:
    input:
        expand(f"{RESULTDIR}/{{aligner}}/sniffles/{{coverage}}/GM24385.vcf", aligner=["minimap2","ngmlr"], coverage=["5X", "10X", "15X", "20X", "25X", "35X", "total"])

rule svim:
    input:
        expand(f"{RESULTDIR}/{{aligner}}/svim/{{coverage}}/GM24385.vcf", aligner=["minimap2","ngmlr"], coverage=["5X", "10X", "15X", "20X", "25X", "35X", "total"])

rule pbsv:
    input:
        expand(f"{RESULTDIR}/pbmm2/pbsv/{{coverage}}/GM24385.vcf",coverage=["5X", "10X", "15X", "20X", "25X", "35X", "total"])

rule npinv:
    input:
        expand(f"{RESULTDIR}/{{aligner}}/npinv/{{coverage}}/GM24385.vcf", aligner=["minimap2","ngmlr"], coverage=["5X", "10X", "15X", "20X", "25X", "35X", "total"])

rule nanosv:
    input:
        expand(f"{RESULTDIR}/{{aligner}}/nanosv/{{coverage}}/GM24385.vcf", aligner=["minimap2","ngmlr"], coverage=["5X", "10X", "15X", "20X", "25X", "35X", "total"])
